tests/basic/replay.rs

```diff
@@ -178,3 +178,115 @@ fn replay_persist_none() {
     // All our current failure persistence modes print the word "schedule", so check that's missing
     assert!(!output.contains("schedule"));
 }
+
+/// Tests that events not causally related to the failure are never scheduled.
+#[test]
+fn replay_causality() {
+    // The main thread will spawn three threads:
+    // - A, which acquires the lock and sets it to one;
+    // - B, which acquires the lock and asserts it is zero;
+    // - C, which sets an unrelated atomic Boolean.
+    // If A runs before B (as in the schedule below), then the assertion
+    // fails. If we provide the clock of this failure to the scheduler, we
+    // should never see thread C do anything, i.e., the atomic Boolean should
+    // never be set, because it is not causally related to the actual panic.
+
+    use std::sync::atomic::{AtomicBool, Ordering};
+
+    let flag = Arc::new(AtomicBool::new(false));
+    let flag_clone = Arc::clone(&flag);
+
+    let result = panic::catch_unwind(|| {
+        let schedule = Schedule::new_from_task_ids(0, vec![0, 1, 0, 2, 0, 3, 0, 1, 1, 2, 2]);
+        let mut scheduler = ReplayScheduler::new_from_schedule(schedule);
+        scheduler.set_target_clock(&[2, 2, 1]);
+        let mut config = Config::new();
+        config.failure_persistence = FailurePersistence::None;
+        let runner = Runner::new(scheduler, config);
+        runner.run(move || {
+            let lock = Arc::new(Mutex::new(0usize));
+            let lock_clone = Arc::clone(&lock);
+            thread::spawn(move || {
```

Comment:
Since we're relying on specific TaskIds here, could you add an `assert_eq!(me(), 1)` etc to each of these threads?